airflow/plugins/operators/scrape_ntd_api.py

```diff
@@ -0,0 +1,148 @@
+    and also streams to console.
+    """
+    logger = logging.getLogger(__name__)
+    logger.setLevel(logging.INFO)
+    formatter = logging.Formatter(
+        "%(asctime)s:%(levelname)s: %(message)s", datefmt="%y-%m-%d %H:%M:%S"
+    )
+    file_handler = logging.FileHandler(logfilename)
+    file_handler.setFormatter(formatter)
+    stream_handler = logging.StreamHandler()
+    stream_handler.setFormatter(formatter)
+
+    if not logger.hasHandlers():
+        logger.addHandler(file_handler)
+        logger.addHandler(stream_handler)
+
+    return logger
+
+
+class NtdDataProductAPIExtract(BaseModel):
+    endpoint_id: str
+    product: str
+    year: str
+    data: Json
+    logger: Optional[logging.Logger]
+    extract_time: Optional[pendulum.DateTime]
+
+    # bucket = CALITP_BUCKET__NTD_DATA_PRODUCTS
+    logger = write_to_log("load_ntd_apidata_output.log")
+    extract_time = pendulum.now()
+
+    # pydantic doesn't know dataframe type
+    # see https://stackoverflow.com/a/69200069
+    class Config:
+        arbitrary_types_allowed = True
+
+    def fetch_from_ntd_api(self):
+        """Restore comments"""
+
+        self.logger.info(f"Downloading NTD data for {self.year} / {self.product}.")
+
+        url = (
+            "https://data.transportation.gov/resource/"
+            + self.endpoint_id
+            + ".json"
+            + "?$limit=5000000"
+        )
+
+        response = requests.get(url).content
+
+        if response is None or len(response) == 0:
+            self.logger.info(
+                f"There is no data to download for {self.year} / {self.product}. Ending pipeline."
+            )
+            pass
+        else:
+            self.data = response
+            self.logger.info(
+                f"Downloaded {self.product} data for {self.year} with {len(self.data)} rows!"
+            )
+
+    def make_hive_path(self, product: str, bucket: str):
+        if not self.extract_time:
+            raise ValueError(
+                "An extract time must be set before a hive path can be generated."
+            )
+        return os.path.join(
+            bucket,
+            f"{self.product}_{self.year}",
+            f"year={self.extract_time.format('YYYY')}",
+            f"dt={self.extract_time.to_date_string()}",
+            f"ts={self.extract_time.to_iso8601_string()}",
+            f"{self.product}_{self.year}.jsonl.gz",
+        )
+
+    def save_to_gcs(self, fs, bucket):
+        hive_path = self.make_hive_path(self.product, bucket)
+        self.logger.info(f"Uploading to GCS at {hive_path}")
+        if (self.data is None) or (len(self.data) == 0):
+            self.logger.info(
+                f"There is no data for {self.year} / {self.product}, not saving anything. Pipeline exiting."
+            )
+            pass
+        else:
+            fs.pipe(hive_path, gzip.compress(self.data))  # .encode()),
+        return hive_path
+
+
+class NtdDataProductAPIOperator(BaseOperator):
```

Comment:
Is there a reason we'd go the custom operator route here instead of, say, a `PythonOperator` pointing to a script? It may be a bias of mine, but I feel like custom operators are best left for operations that are going to be used in multiple places.

On that note, have we considered doing a separate task for each NTD API endpoint? E.g., have one `NtdDataProductEndpointOperator` or something, and use that once for each endpoint. I don't have context for how the NTD API might fail, and I could see it being frustrating to have the whole task fail because one endpoint failed.